04/src/academy/pocu/comp2500samples/w04/singletonlogger/Logger.java

```diff
@@ -0,0 +1,110 @@
+
+    private static Logger instance;
+
+    private LogLevel minLogLevel;
+    private String outputLogFilePath;
+
+    private Logger(LogLevel minLogLevel, String outputLogFilePath) {
+        this.minLogLevel = minLogLevel;
+        this.outputLogFilePath = outputLogFilePath;
+    }
+
+    public static Logger getInstance() throws IOException {
+        if (instance == null) {
+            String classPath = getClassPath();
+            Path loggerConfigPath = Paths.get(classPath, CONFIG_FILE_NAME);
+
+            File configFile = new File(loggerConfigPath.toString());
+
+            LogLevel defaultLogLevel = LogLevel.WARNING;
+            String defaultOutputFilename = "log.txt";
+
+            if (configFile.isFile()) {
+                List<String> lines = Files.readAllLines(loggerConfigPath, StandardCharsets.UTF_8);
+
+                for (String line : lines) {
+                    String[] splits = line.split("=");
+
+                    switch (splits[0]) {
+                        case "loglevel":
+                            defaultLogLevel = LogLevel.valueOf(splits[1]);
+                            break;
+
+                        case "output":
+                            defaultOutputFilename = splits[1];
+                            break;
+
+                        default:
+                            throw new IllegalArgumentException("Unknown configuration setting: " + splits[0]);
+                    }
+                }
+            }
+
+            Path path = Paths.get(classPath, defaultOutputFilename);
+            String pathString = path.toString();
+
+            BufferedWriter out = new BufferedWriter(new FileWriter(pathString));
+            out.close();
+
+            instance = new Logger(defaultLogLevel, pathString);
+        }
+
+        return instance;
+    }
+
+    public void logDebug(String message, Object ... args) throws IOException {
+        writeToFile(LogLevel.DEBUG, message, args);
+    }
+
+    public void logInformation(String message, Object ... args) throws IOException {
+        writeToFile(LogLevel.INFORMATION, message, args);
+    }
+
+    public void logWarning(String message, Object ... args) throws IOException {
+        writeToFile(LogLevel.WARNING, message, args);
+    }
+
+    public void logError(String message, Object ... args) throws IOException {
+        writeToFile(LogLevel.ERROR, message, args);
+    }
+
+    public void logCritical(String message, Object ... args) throws IOException {
+        writeToFile(LogLevel.CRITICAL, message, args);
+    }
+
+    private void writeToFile(LogLevel logLevel, String message, Object ... args) throws IOException {
+        if (this.minLogLevel.getLogLevel() <= logLevel.getLogLevel()) {
+            String log = String.format("[%s] %s: %s", Instant.now().toString(), logLevel.toString(), String.format(message, args));
+
+            BufferedWriter out = new BufferedWriter(new FileWriter(this.outputLogFilePath, true));

```

Comment:
honestly i think u would have bufferedwriter open all the time in Logger class.. and keep writing instead of opening it all the time.

Reply:
but then I would actually need createInstance() and disposeInstance() to open and close BufferedWriter. If you are fine with it, then I can do that instead.

Reply:
why can't you do taht in constructor?

Reply:
I can open in the constructor. Where do I close the buffer then? A separate dispose method?

Reply:
why do u ever need it... I believe StreamWriter will close it when it's GCed? lol

Reply:
I think you can just create deleteInstance() method to delete it manually. then later when someone calls getInstance() it'll create it again right? so it's safe.. and u can do early release